File: Queue/232-implement-queue-using-stacks.py
"""
232. 用栈实现队列
简单
https://leetcode-cn.com/problems/implement-queue-using-stacks/
"""


# 36 ms	    85.00%
# 13.7 MB   11.11%
class MyQueue:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.inStack = []
        self.outStack = []


    def push(self, x: int) -> None:
        """
        Push element x to the back of queue.
        """
        self.inStack.append(x)


    def pop(self) -> int:
        """
        Removes the element from in front of queue and returns that element.
        """
        if len(self.outStack):
            return self.outStack.pop(0)
        else:
            while len(self.inStack):
                self.outStack.append(self.inStack.pop(0))
            return self.outStack.pop(0)

    def peek(self) -> int:
        """
        Get the front element.
        """
        if len(self.outStack):
            return self.outStack[0]
        else:
            while len(self.inStack):
                self.outStack.append(self.inStack.pop(0))
            return self.outStack[0]

    def empty(self) -> bool:
        """
        Returns whether the queue is empty.
        """
        return len(self.inStack) == 0 and len(self.outStack) == 0



# Your MyQueue object will be instantiated and called as such:
obj = MyQueue()
# obj.push(1)
# obj.push(2)
# param_2 = obj.pop()
# param_3 = obj.peek()
print(obj.empty())

File: Leetcode-Swift/Leetcode/232.implement-queue-using-stacks.swift
//
//  232.implement-queue-using-stacks.swift
//  Leetcode-Swift
//
//  Created by devcz on 2021/4/4.
//

import Foundation
/*
 232. 用栈实现队列 (简单)
 https://leetcode-cn.com/problems/implement-queue-using-stacks/
 */

class MyQueue232 {
    var inputStack: [Int] = []
    var outputStack: [Int] = []
    
    /** Initialize your data structure here. */
    init() {
        
    }
    
    /** Push element x to the back of queue. */
    func push(_ x: Int) {
        inputStack.append(x)
    }
    
    /** Removes the element from in front of queue and returns that element. */
    func pop() -> Int {
        if outputStack.count == 0 {
            while inputStack.count > 0 {
                outputStack.append(inputStack.popLast() ?? 0)
            }
        }
        return outputStack.popLast() ?? 0
    }
    
    /** Get the front element. */
    func peek() -> Int {
        if outputStack.count == 0 {
            while inputStack.count > 0 {
                outputStack.append(inputStack.popLast() ?? 0)
            }
        }
        return outputStack.last ?? 0
    }
    
    /** Returns whether the queue is empty. */
    func empty() -> Bool {
        return inputStack.count == 0 && outputStack.count == 0
    }
    
    func size() -> Int {
        return inputStack.count + outputStack.count
    }
}

class Solution_232 {
    func test() {
        let obj = MyQueue232()
        obj.push(1)
        let ret_2: Int = obj.pop()
        obj.push(2)
        let ret_3: Int = obj.peek()
        let ret_4: Int = obj.pop()
        let ret_5: Bool = obj.empty()
        print(ret_2, ret_3, ret_4, ret_5)
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * let obj = MyQueue()
 * obj.push(x)
 * let ret_2: Int = obj.pop()
 * let ret_3: Int = obj.peek()
 * let ret_4: Bool = obj.empty()
 */
